src/Mono.Android/Java.Lang/Object.cs

```diff
@@ -295,7 +310,9 @@ protected void SetHandle (IntPtr value, JniHandleOwnership transfer)
 				return r;
 			}
 
-			return Java.Interop.TypeManager.CreateInstance (handle, transfer, type);
+			JniObjectReference reference = JNIEnv.CreateJniObjectReference (handle, transfer);
+			JniObjectReferenceOptions options = JNIEnv.ToJniObjectReferenceOptions (transfer);
+			return (IJavaPeerable) JNIEnvInit.AndroidValueManager?.GetValue (ref reference, options, type);
```

Comment:
I was wrong about always wanting this cast.

That said, IIRC a previous version of this PR instead used `as IJavaPeerable`, which also seems wrong.

The problem: [`Java.InteropTests.JnienvTest.MoarThreadingTests` is failing](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=10731022&view=ms.vss-test-web.build-test-results-tab&runId=118389860&resultId=100110&paneView=debug):

```
No exception should be thrown [t2]! Got: System.InvalidCastException: Arg_InvalidCastException
at Java.Lang.Object.GetObject(IntPtr , JniHandleOwnership , Type )
at Android.Runtime.JNIEnv.<CreateNativeArrayElementToManaged>g__GetObject|74_11(IntPtr , Type )
at Android.Runtime.JNIEnv.<>c.<CreateNativeArrayElementToManaged>b__74_9(Type type, IntPtr source, Int32 index)
at Android.Runtime.JNIEnv.GetObjectArray(IntPtr , Type[] )
at Java.InteropTests.JnienvTest.<>c__DisplayClass26_0.<MoarThreadingTests>b__1()
Expected: null
But was:  <System.InvalidCastException: Arg_InvalidCastException
at Java.Lang.Object.GetObject(IntPtr , JniHandleOwnership , Type )
at Android.Runtime.JNIEnv.<CreateNativeArrayElementToManaged>g__GetObject|74_11(IntPtr , Type )
at Android.Runtime.JNIEnv.<>c.<CreateNativeArrayElementToManaged>b__74_9(Type type, IntPtr source, Int32 index)
at Android.Runtime.JNIEnv.GetObjectArray(IntPtr , Type[] )
at Java.InteropTests.JnienvTest.<>c__DisplayClass26_0.<MoarThreadingTests>b__1()>
```

https://github.com/dotnet/android/blob/8d7713071c40766651664de25b53eb15225785d0/tests/Mono.Android-Tests/Java.Interop/JnienvTest.cs#L363-L386

`JnienvTest.cs:386` is the source of the `JNIEnv.GetObjectArray()` invocation at the root of the stack trace.

The cast on this line is the source of the `InvalidCastException`.

The question is: how did this previously work, and why is it failing now?

Tracing through how it previously worked:

 1. [`JNIEnv.GetObjectArray()`](https://github.com/dotnet/android/blob/8d7713071c40766651664de25b53eb15225785d0/src/Mono.Android/Android.Runtime/JNIEnv.cs#L1096-L1121) is called; `element_types=new[]{typeof(int)}`.
 2. `JNIEnv.GetObjectArray()` calls [`GetConverter<T>(NativeArrayElementToManaged, null, array_ptr)`](https://github.com/dotnet/android/blob/8d7713071c40766651664de25b53eb15225785d0/src/Mono.Android/Android.Runtime/JNIEnv.cs#L716)
 3. `GetConverter<T>()` sees that `array_ptr`/`array` is not null; calls `GetClassNameFromInstance(array)`.
 4. `GetClassNameFromInstance(array)` *should* be `[Ljava/lang/Object;`, because the array was created via `JNIEnv.NewObjectArray<int> (new[]{1})`.
 5. `GetConverter<T>()` thus skips the `switch` on lines 732-749
 6. `GetConvter<T>()` should thus return a `Func<Type?, IntPtr, int, object?>` for `IJavaObject`: https://github.com/dotnet/android/blob/8d7713071c40766651664de25b53eb15225785d0/src/Mono.Android/Android.Runtime/JNIEnv.cs#L703-L708
 7. Returning to `GetObjectArray()`, line 1110 `targetType` will be `typeof(int)`, we call `converter(null, array_ptr, i)`.
 8. `value` should thus be a `java.lang.Integer`/`Java.Lang.Integer` instance, and the `Convert.ChangeType()` on line should turn it into an `int`, as `Integer` implements `IConvertible`.

So why's this change fail?  *Probably* because `AndroidValueManager?.GetValue(…)` has different semantics from `TypeManager.CreateInstance(…)`.  I need to further investigate.

Reply:
@jonpryor [wrote](https://github.com/dotnet/android/pull/9630/files#r1891003136):

> So why's this change fail?  *Probably* because `AndroidValueManager?.GetValue(…)` has different semantics from `TypeManager.CreateInstance(…)`.  I need to further investigate.

Time for further investigation!  Consider the following patch to java-interop/samples/Hello-Java.Base:

```diff
diff --git a/samples/Hello-Java.Base/Program.cs b/samples/Hello-Java.Base/Program.cs
index b1e0d5ba..9e63f3ba 100644
--- a/samples/Hello-Java.Base/Program.cs
+++ b/samples/Hello-Java.Base/Program.cs
@@ -61,10 +61,15 @@ namespace Hello
 			CreateJLO ();
 		}
 
-		static void CreateJLO ()
+		static unsafe void CreateJLO ()
 		{
-			var jlo = new Java.Lang.Object ();
-			Console.WriteLine ($"binding? {jlo.ToString ()}");
+			var i_class = new JniType ("java/lang/Integer");
+			var i_ctor  = i_class.GetConstructor ("(I)V");
+			JniArgumentValue* i_args = stackalloc JniArgumentValue [1];
+			i_args [0] = new JniArgumentValue (42);
+			var i_value = i_class.NewObject (i_ctor, i_args);
+			var v = JniEnvironment.Runtime.ValueManager.GetValue (ref i_value, JniObjectReferenceOptions.CopyAndDispose, null);
+			Console.WriteLine ($"v? {v} {v?.GetType ()}");
 		}
 
 		static void ReportTiming ()
```

This creates a new `java.lang.Integer(42)`, invokes `Runtime.ValueManager.GetValue()`, and prints the resulting value and its type.

The output is:

```dotnetcli
% ~/Downloads/dotnet-sdk-8.0.206-osx-x64/dotnet run --project samples/Hello-Java.Base/*.csproj
…
v? 42 System.Int32
```

(I have to use .NET 8.0.206 because anything later crashes.  Hopefully this will be fixed in the January .NET 8 service release…)

Note that `Runtime.GetValue(objref, …, null)` converts to a `System.Int32`, *not* a `Java.Lang.Integer`.  *This* is why there's an `InvalidCastException`.

Reply:
Now that the [cause of the `InvalidCastException` is understood](https://github.com/dotnet/android/pull/9630/files#r1891086768), what do we do about it?

There are at least two paths worth pursuing:

 1. Don't Do That™ (don't use `ValueManager.GetValue()` here)
 2. Don't Do That™ (update `ValueManager.GetValue()` to not auto-unbox like that.)

(1) should be possible, by e.g. updating `TypeManager.CreateInstance()` to use methods on `JniEnvironment.Runtime.TypeManager` instead of using the P/Invokes within `GetClassName()`/etc.

(2) should *also* be possible, but may be more time consuming, depending on how many unit tests it breaks.